src/index.ts

```diff
@@ -1,114 +1,121 @@
     log('Weekend updates not enabled, skipping')
-    await clearSlackStatus(cachedTrack)
+    await clearOnEnded(cachedTrack)
     return
   }
 
-  // Status restrictions
-  const currentPresence = await getSlackPresence()
-  if (currentPresence === 'away') {
-    log('User presence is "away"')
-    await clearSlackStatus(cachedTrack)
-    return
-  }
-
-  const currentProfile = await getSlackProfile()
-  if (!shouldSetStatus(currentProfile)) {
-    log('Custom status detected')
-    return
-  }
-
-  // Now playing restrictions
-  const recentTracks = await getRecentLastFmTracks(config.lastFM.username)
-  const nowPlaying = getNowPlaying(recentTracks.track)
+  // Get recent track
+  // ---
+  const tracks = await lastFm.getRecentTracks()
+  const nowPlaying = tracks.find(t => t.nowPlaying)
 
   if (nowPlaying === undefined) {
-    log('Nothing playing')
-    await clearSlackStatus(cachedTrack)
-    await deleteTrackJSON()
+    log('Nothing playing', 'bot')
+    await Cache.deleteTrackJSON()
+    await clearOnEnded(cachedTrack)
     return
   }
 
-  // Equality restriction, don't update if it's not necessary
-  if (trackIsEqual(nowPlaying, cachedTrack)) {
-    log('Now playing track is cached, no update necessary')
+  if (
+    cachedTrack?.name === nowPlaying?.name &&
+    cachedTrack?.artist === nowPlaying?.artist
+  ) {
+    log('Now playing track is cached, no update necessary', 'bot')
     return
   }
 
-  const track = await getLastFmTrack(nowPlaying.name, nowPlaying.artist['#text'])
-  let duration = 60 * (config.updateExpiration * 1000)
-  let status = `${nowPlaying.name} ${config.slack.separator} ${nowPlaying.artist['#text']}`
+  // Attempt to find more details (duration)
+  // ---
+  const track = await lastFm.getTrack(nowPlaying.name, nowPlaying.artist)
 
-  if (track !== undefined) {
-    await writeTrackJSON(track)
+  // Compute new status and duration
+  // ---
+  const trackDuration = track?.duration ?? 60 * (10 * 1000) // 10 minutes
+  const trackName = track?.name ?? nowPlaying.name
+  const trackArtist = track?.artist ?? nowPlaying.artist
+
+  const status = `${trackName} ${config.app.separator} ${trackArtist}`
 
-    status = `${track.name} ${config.slack.separator} ${track.artist.name}`
-    duration = track.duration !== '0' ? Number(track.duration) : duration
+  if (track !== undefined) {
+    await Cache.writeTrackJSON(track)
   } else {
-    log('No detailed track info found, falling back to recent track', 'lastfm')
-    await deleteTrackJSON()
+    await Cache.deleteTrackJSON()
   }
 
-  log(`Setting status to "${status}"`, 'slack')
-  await setSlackStatus(status, duration)
-}
+  // Propagate updates to all Slack Teams
+  // ---
+  for (const client of clients) {
+    const [presence, profile] = await Promise.all([
+      client.getPresence(),
+      client.getProfile()
+    ])
+
+    if (presence !== 'active') return
+    if (!profile.status_text.includes(config.app.separator)) return
```

Comment:
Not necessary to batch these request, we can avoid the profile request if the presence is `away`.